lib/malan/accounts/session.ex

```diff
@@ -19,11 +19,28 @@ defmodule Malan.Accounts.Session do
     field :valid_only_for_ip, :boolean, default: false
     # This is not yet used but will be in #78
     field :valid_only_for_approved_ips, :boolean, default: false
+    # Absolute limit of extension.  Can never extend session beyond this point
+    field :extendable_until, :utc_datetime
+    # Longest a session can be extended for on each extension request
+    field :max_extension_secs, :integer
```

Comment:
It's probably overkill, but my initial thought was basically that during authentication (when the username and password are proven) the session is the most trustworthy, and after that if an API token were leaked or compromised it would at least provide an upper bound on how long it could be used (i.e. not used to extend the session infinitely), giving the client options.  When first authenticating the session, the authenticator can (optionally) specify the session length.  Example (hypothetical) scenario:  User A authenticates, setting an expiration of 1 week and doesn't want the token to have an infinite life span, but does want to be able to renew it a few times up to a maximum date (say 4 weeks total).  After authenticating, if this option weren't present, the token could be immediately extended to 4 weeks without requiring use.  It's definitely a small thing and maybe not worth the added complexity, but it's a way to gate the extension length (so for example any token is only valid for ). 

The other scenario I considered was an "active" session extension vs stale extension implemented by a client.  For example if the client wants to have sessions expire automatically in they are "inactive" for some period of time, then this could be a useful defense-in-depth as it can be used to limit how far in the future a session can be valid without explicit extending, while still allowing a longer life of the token before hard expiration.  Example:  Client wants tokens to expire after 30 minutes if not renewed (which renewal happens automatically when they use the app).  On the client-side you can then renew the token each time it's needed when the app is being used.  If the app isn't used, the token will automatically expire and require reauthentication.

If the authenticator doesn't care about this, they can also leave it empty as it will default to the same length as the original lifestpan of the token.

Now thinking more through that previous scenario, that could easily result in a *ton* of "extensions" which would make the need for paginating the API a lot higher...

Gonna need to think on this for a bit :thinking:

Reply:
One of the scenarios I wanted to make reasonable in the future also was "option A" which was basically token automatically extends whenever it's used.  Thinking more though that would result in huge numbers of these extension records which would definitely become a performance problem. 

How important do you think having the record is?  Explicit extensions feel important to have records for, but how important would it be to track automatic extension events?  And if we tracked explicit ones but not automatic ones, would that be unexpected for the client? 

We could just split the extensions into a proper table.  That was something I debated initially and decided against to prevent the scope from getting even bigger, but if we need to track every extension (even automatic) then we probably should split it and that would be much easier to do now before it goes into production

Reply:
I agree, it's definitely helpful to finalize the data model before it goes out!

For comparison, I'm thinking about applications that offer a visible security log. I can't recall seeing any that enumerated every time my session refreshed, and I'm sure at least one of those was an OAuth setup with refresh tokens and access tokens. So I don't have a good sense of the use case here from a user-facing security log perspective.

That leaves me considering some potential forensic utility. I could imagine wanting to report on what IP addresses refreshed a token. That would be a very rare need, though, and I wouldn't be surprised if I needed special tools to access such a report.

I would expect such a log, if it existed, to contain implicit extensions as well as explicit extensions. That would be forensically useful to eliminate certain degenerate or malicious scenarios.